components/validator/guardian.go

```diff
@@ -46,59 +67,289 @@ func NewGuardian(cfg Config, l log.Logger) (*Guardian, error) {
+	}
+
+	if g.deletionRequestedSub != nil {
+		g.deletionRequestedSub.Unsubscribe()
 	}
 
 	g.cancel()
 	g.wg.Wait()
+
 	close(g.validationRequestedChan)
+	close(g.deletionRequestedChan)
 
 	return nil
 }
 
-func (g *Guardian) handleValidationRequested() {
+func (g *Guardian) initSub() {
+	opts := utils.NewSimpleWatchOpts(g.ctx)
+
+	g.validationRequestedChan = make(chan *bindings.SecurityCouncilValidationRequested)
+	g.validationRequestedSub = event.ResubscribeErr(time.Second*10, func(ctx context.Context, err error) (event.Subscription, error) {
+		if err != nil {
+			g.log.Warn("resubscribing after failed ValidationRequested event", "err", err)
+		}
+		return g.securityCouncilContract.WatchValidationRequested(opts, g.validationRequestedChan, nil)
+	})
+
+	g.deletionRequestedChan = make(chan *bindings.SecurityCouncilDeletionRequested)
+	g.deletionRequestedSub = event.ResubscribeErr(time.Second*10, func(ctx context.Context, err error) (event.Subscription, error) {
+		if err != nil {
+			g.log.Warn("resubscribing after failed DeletionRequested event", "err", err)
+		}
+		return g.securityCouncilContract.WatchDeletionRequested(opts, g.deletionRequestedChan, nil, nil)
+	})
+}
+
+// inspectorLoop finds and deletes outputs whose zk fault proving has failed due to an undeniable bug
+// among whose creation period has passed but not finalized
+func (g *Guardian) inspectorLoop() {
+	defer g.wg.Done()
+
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+
+	for ; ; <-ticker.C {
+		select {
+		case <-g.ctx.Done():
+			return
+		default:
+			status, err := g.cfg.RollupClient.SyncStatus(g.ctx)
+			if err != nil {
+				g.log.Error("guardian: failed to get sync status", "err", err)
+				continue
+			}
+
+			var currentL2 *big.Int
+			if g.cfg.AllowNonFinalized {
+				currentL2 = new(big.Int).SetUint64(status.SafeL2.Number)
+			} else {
+				currentL2 = new(big.Int).SetUint64(status.FinalizedL2.Number)
+			}
+
+			latestOutputL2, err := g.fetchLatestOutputL2Block(g.ctx)
+			if latestOutputL2.Cmp(currentL2) == 1 {
+				g.log.Info("guardian: need to wait for L2 blocks sync completed", "currentL2", currentL2, "latestOutputL2", latestOutputL2)
+				continue
+			}
+
+			// currentL1 and creationEndedL1 are used for searching events of ReadyToProve in L1 blocks
+			currentL1 := new(big.Int).SetUint64(status.CurrentL1.Number)
+			creationPeriodL1 := new(big.Int).Div(g.creationPeriodSeconds, big.NewInt(12))
+			creationPeriodL2 := new(big.Int).Div(g.creationPeriodSeconds, g.l2BlockTime)
+			if currentL1.Cmp(creationPeriodL1) != 1 || currentL2.Cmp(creationPeriodL2) != 1 {
+				g.log.Warn("guardian: there is no output when the creation period is over yet", "currentL1", currentL1, "currentL2", currentL2, "creationPeriodL1", creationPeriodL1, "creationPeriodL2", creationPeriodL2)
+				continue
+			}
+
+			creationEndedL1 := new(big.Int).Sub(currentL1, creationPeriodL1)
+			g.updateInspectionWindow(creationEndedL1, currentL1)
```

Comment:
I wonder it does work fine.. here we update inspection window using current L1 info, and below in `inspectOutput` goroutine we use the updated inspection window for all outputs in checkpoint range.
1. is it ok to use the same inspection window for all outputs in checkpoint range?
2. After sending the goroutines and loop back here, update the inspection window again, won't this affect the inspection window value already used in the sent goroutines?

Reply:
Thank you for pointing this out!
As you said, different inspection window should be used for each output.
I made a change for this.